# Урок 1
## Случайные события. Условная вероятность. Формула Байеса. Независимые испытания
### Домашнее задание

In [1]:
import numpy as np


def arrangements(n: int, k: int) -> int:
    """Число размещений.
    """
    
    return np.math.factorial(n) // np.math.factorial(n - k)


def permutations(n: int) -> int:
    """Число перестановок.
    """
    
    return np.math.factorial(n)


def combinations(n: int, k: int) -> int:
    """Число сочетаний.
    """
    
    return np.math.factorial(n) // (np.math.factorial(k) * np.math.factorial(n - k))

### Задача 1
Из колоды в 52 карты вынимают случайным образом 4 карты. Найти число исходов, соответствующих тому, что был вытянут хотя бы один туз.

`a` - число исходов, когда из колоды без тузов выбраны 4 карты

`b` - число исходов, когда вытянуты 4 карты из полной колоды

`b-a` - число исходов, соответствующих тому, что был вытянут хотя бы один туз.

In [2]:
a = combinations(48,4)
b = combinations(52,4)
b-a

76145

#### Ответ: 76145 исходов

### Задача 2
Семь человек рассаживаются наудачу на скамейке. Какова вероятность того, что рядом будут сидеть: 

а) два определённых человека? 

б) три определённых человека?

а) Всего существует 6 пар соседних мест, итого 2 определенных человека могут сидеть рядом дург с другом 12-ю способами. Количество всех возможных рассадок людей на лавке - размещения 7 элементов.

Соответственно вероятность:

In [3]:
12/permutations(7)

0.002380952380952381

б) По аналогии с примером (а) всего существует 5 троек соседних мест, итого 3 определенных человека могут сидеть рядом дург с другом количеством способов, определяемым как размещение трех элементов умноженное на 5. Количество всех возможных рассадок людей на лавке - размещения 7 элементов.

Соответственно вероятность:

In [4]:
5*permutations(3)/permutations(7)

0.005952380952380952

### Задача 3
Из 60 вопросов, входящих в экзаменационные билеты, студент знает 50. 

Какова вероятность того, что среди трёх наугад выбранных вопросов студент знает: 

а) все? 

б) два?

а)

Количество исходов, когда все выбраные вопросы студент знает - сочетание из 3 по 50.

Количество всех возможных исходов выбора трех вопросов - сочетание из 3 по 60.

Вероятность:

In [5]:
combinations(50,3)/combinations(60,3)

0.5727644652250146

Или произведение вероятностей последовательного вытягивания трех правильных вопросов.

In [6]:
(50/60)*(49/59)*(48/58)

0.5727644652250146

б)

Количество исходов, когда вытянуты два известных вопроса - сочетание 2 из 50.

Количество исходов, когда вытянут один неизвестный вопрос - сочетание 1 из 10 (очевидно 10)

Количество исходов, когда из трех вытянуты два известных - перемножение вышеприведенных данных, то есть сочетание 2 из 50 умножить на 10

Далее получаем вероятность по аналогии с примером (а)

In [7]:
combinations(50,2)*10/combinations(60,3)

0.35797779076563413

### Задача 4
Бросается игральная кость. Пусть событие A - появление чётного числа, событие B - появление числа больше трёх. Являются ли эти события независимыми?

P(A) = P(B) = 1/2

P(AB) = 2/6 = 1/3

P(A|B) = P(AB)/P(A) = (1/3) / (1/2) = 2/3

P(A|B) не равно P(A), значит A и B - зависимые события. Действительно, если известно, что выпало число больше трех, значит это 4,5 или 6. Из этой тройки чисел - два нечетны, и только одно четно. То есть при условии B вероятность выпадения четного повышается.

### Задача 5
Допустим, имеется некоторая очень редкая болезнь (поражает 0.1% населения). Вы приходите к врачу, вам делают тест на эту болезнь, и тест оказывается положительным. Врач говорит вам, что этот тест верно выявляет 99% больных этой болезнью и всего лишь в 1% случаев даёт ложный положительный ответ.
Вопрос: какова вероятность, что вы действительно больны ей?
Подсказка: вновь используйте формулу Байеса с раскрытием знаменателя с помощью формулы полной вероятности.

Дано: 

A - результат теста положителен

В - болен

Р(A|B) = 0.99 (тест положителен, при условии реального нличия болезни)

P(A|~B) = 0.01 (тест положителен, при условии отсутствия болезни)

P(B) = 0.001

In [8]:
P_A_if_B = 0.99
P_A_if_not_B = 0.01
P_B = 0.001

По формуле полной вероятности: P(A) = P(A|B) * P(B) + P(A|~B) * (1 - P(B))

In [9]:
P_A = P_A_if_B*P_B + P_A_if_not_B*(1-P_B)
P_A

0.01098

По формуле байеса: P(B|A) = P(A|B) * P(B) / P(A)

In [10]:
P_B_if_A = P_A_if_B * P_B / P_A
P_B_if_A

0.09016393442622951

#### Ответ: Вероятность того, что пациент действительно болен при условии, что тест положителен, 9%